rust/gui-client/src-tauri/src/client/updates.rs

```diff
@@ -1,18 +1,199 @@
+    WaitRandom,
+    /// Check the latest version from the Firezone website and write it to disk.
+    CheckNetwork,
+    /// Wait approximately a day using `tokio::time::interval`.
+    WaitInterval,
+}
+
+impl Checker {
+    fn new(ours: Version) -> Self {
+        Self {
+            latest_seen: None,
+            ours,
+            state: CheckerStep::CheckFile,
+        }
+    }
+
+    /// Call this when we just checked the network or the file
+    ///
+    /// Returns the release if we should alert the UI about this release
+    #[must_use]
+    fn handle_check(&mut self, release: Release) -> Option<Release> {
+        match self.state {
+            // After startup and checking the file, wait a random amount of time.
+            CheckerStep::CheckFile => self.state = CheckerStep::WaitRandom,
+            // Always wait a full interval after a network check
+            CheckerStep::CheckNetwork => self.state = CheckerStep::WaitInterval,
+            // If we weren't waiting on an I/O check, something is wrong
+            CheckerStep::WaitRandom | CheckerStep::WaitInterval => {
+                panic!("Impossible, got `handle_check` when update checker was waiting for wakeup")
+            }
+        }
+
+        let newer_than_ours = release.version > self.ours;
+        let newer_than_latest_seen = match &self.latest_seen {
+            None => true,
+            Some(latest_seen) => release.version > *latest_seen,
+        };
+        self.latest_seen = Some(release.version.clone());
+        if newer_than_ours && newer_than_latest_seen {
+            Some(release)
+        } else {
+            None
+        }
+    }
+
+    /// Call this when the file is missing or invalid, or the network response is invalid or the network is unreachable.
+    fn handle_failed_check(&mut self) {
+        match self.state {
+            // After startup and checking the file, wait a random amount of time.
+            CheckerStep::CheckFile => self.state = CheckerStep::WaitRandom,
+            // Always wait a full interval after a network check
+            CheckerStep::CheckNetwork => self.state = CheckerStep::WaitInterval,
+            // If we weren't waiting on an I/O check, something is wrong
+            CheckerStep::WaitRandom | CheckerStep::WaitInterval => {
+                panic!("Impossible, got `handle_failed_check` when update checker was waiting for wakeup")
+            }
+        }
+    }
+
+    /// Call this when we just woke up from sleeping
+    fn handle_wake(&mut self) {
+        match self.state {
+            CheckerStep::CheckFile | CheckerStep::CheckNetwork => {
+                panic!("Impossible, got `handle_wake` when update checker was waiting for I/O")
+            }
+            CheckerStep::WaitRandom | CheckerStep::WaitInterval => {
+                self.state = CheckerStep::CheckNetwork
+            }
+        }
+    }
+
+    fn poll(&self) -> CheckerStep {
```

Comment:
I'd expect `poll` to make progress and thus require `&mut self`.

```suggestion
    fn state(&self) -> CheckerStep {
```